File: algorithms/xlm/xlm.py
"""
extreme learning machine in python:
x1 x2 x3 x4
-   -   -   -   h = G(w*x + b)
h1  h2  h3
-   -   -       o = h^-1 * y  ???
o
"""
import numpy as np

class XLM:
    def __init__(self, no_hidden_units):
        self.no_hidden_units = no_hidden_units

    def fit(self, X, labels):
        """
        X is a 1-d array of inputs,
        labels are a 1-d array of outputs
        """
        # add a column of all ones to represent the biases
        X = np.column_stack([X, np.ones([X.shape[0]])])
        # weights and baises are all random numbers
        self.random_weights = np.random.randn(X.shape[1], self.no_hidden_units)
        # G = activation_function(weights * inputs + biases)
        G = self.activation_function(X.dot(self.random_weights))
        # the final layer is not random 
        self.w = np.linalg.pinv(G).dot(labels)

    def predict(self, X):
        X = np.column_stack([X, np.ones([X.shape[0], 1])])
        G = self.activation_function(X.dot(self.random_weights))
        return G.dot(self.w)

    def __str__(self):
        return str(self.w)

    def activation_function(self, x):
        return self.softmax(x)

    def tanh(self, x):
        return np.tanh(x)

    def relu(self, x):
        return np.maximum(0, x)

    def linear(self, x):
        return x

    def softmax(self, x):
        """Compute softmax values for each sets of scores in x."""
        return np.exp(x) / np.sum(np.exp(x), axis=0)

File: algorithms/xlm/xlm.cpp
/*
@author: John DiMatteo
creation date: 03-14-2020
desc:
extreme learning machine using Pytorch Tensors

INPUT
    data        - path to csv, first row header, first column has labels
    numhidden   - number of hidden neurons

OUTPUT
    scores      - Tensor of predictions
*/

#include <torch/torch.h>
#include <iostream>

struct XLM {

    torch::Tensor in_weights;
    torch::Tensor out_weights;

    XLM(int input_size, int no_hidden_units) {
        in_weights = torch::rand({input_size,no_hidden_units});
    }

    void train(torch::Tensor X, torch::Tensor y) {
        std::cout << in_weights << std::endl;
        // dot product between X and the random first layer
        X = torch::matmul(X, in_weights);
        // activation function
        X = torch::relu(X);
        // transpose
        torch::Tensor X_T = torch::t(X);
        out_weights = torch::matmul(
            torch::pinverse(
                torch::matmul(X_T, X)
            ),
            torch::matmul(
                X_T,
                y
            )
        );
    };

    torch::Tensor predict(torch::Tensor X) {
        return torch::matmul(
            torch::relu(torch::matmul(X, in_weights)),
            out_weights
        );
    }
};

int main(){

    int row_size = 8;
    int col_size = 8;
    // int output_size = 1;
    int no_hidden_units = 4;

    auto X = torch::rand({row_size, col_size});
    auto y = torch::randint(0, 2, row_size);

    XLM model(col_size, no_hidden_units);
    model.train(X, y);

    torch::Tensor pred = model.predict(X);

    pred = torch::round(pred);

    std::cout << "Accuracy " << (1 - (torch::sum(pred - y) / row_size)) * 100 << "%" << std::endl;

    return 0;

}